baseplate/server/queue_consumer.py

```diff
@@ -0,0 +1,268 @@
+import abc
+import logging
+import os
+import queue
+import signal
+import socket
+import time
+import uuid
+
+from threading import Thread
+from typing import Any
+from typing import Dict
+from typing import Sequence
+
+from gevent.server import StreamServer
+
+import baseplate.lib.config
+
+logger = logging.getLogger(__name__)
+
+
+class PumpWorker(abc.ABC):
+    """Reads messages off of a message queue and puts them into a queue.Queue for handling by a  MessageHandler.
+
+    The QueueConsumerServer will run a single PumpWorker in it's own thread.
+    """
+
+    @abc.abstractmethod
+    def run(self) -> None:
+        """Run the worker.
+
+        This is the method called by the the PumpWorker Thread within the
+        QueueConsumerServer and is ultimately responsible for getting messages
+        off of the message queue and putting them on the work queue.
+        """
+
+    @abc.abstractmethod
+    def stop(self) -> None:
+        """Signal the PumpWorker that it should stop receiving new messages from it's message queue."""
+
+
+class MessageHandler(abc.ABC):
+    """Processes the messages supplied by the PumpWorker.
+
+    A MessageHandler is expected to have a Baseplate object that it can use to
+    generate a RequestContext and a ServerSpan before processing each message.
+
+    The QueueConsumerServer will run multiple MessageHandlers, each in their own
+    Thread based on the max_concurrency value it is given.  Each MessageHandler
+    will be wrapped in a QueueConsumer object that will handle reading new
+    messages off of the work queue and passing them to the MessageHandler as well
+    as providing the methods that start and stop the work done by the Threads.
+    """
+
+    @abc.abstractmethod
+    def handle(self, message: Any) -> None:
+        """Handle a message supplied by a PumpWorker.
+
+        This method is what the server calls when it reads a new message off of
+        the internal work queue that is fed by the PumpWorker.  This is both the
+        method where you should handle processing the message and setting up
+        the RequestContext and ServerSpan for baseplate.
+
+        While not required, it is reccomended that allow the caller to pass in the
```

Comment:
recommended to allow ?

Reply:
Looking at `KombuMessageHandler` it appears that the caller passes the function to process the message in the constructor, is that right? That all makes sense to me, but the description here makes it seem like that function would be an argument to `handle()` (rather than writing implementations/subclasses of `MessageHandler.handle()` that do the message processing)

Reply:
> it appears that the caller passes the function to process the message in the constructor, is that right

Correct, I'll touch up the language here.  All it's trying to express is that you should do what `KombuMessageHandler` does, not implement the logic to actually process the message directly in the `handle` method of the `MessageHandler` subclass you implement.